### DISCUSSION TITLE
How to put .ts files into folders?

### DISCUSSION THREAD

**Discussion by linfan68** [Moddable-OpenSource/moddable#1116] [Q&A]
Sorry about the dumb question...
I'm planning to write a big Mod with lots of source files using typescript, and these are my files:

```
src
  |--main.ts
  |--folderA
  |    |-- compA. ts
  |
  |-- compB.ts

manifest.json
tsconfig.json
```

in my main.ts:

```ts
import CompA from './folderA/compA.ts'
import CompB from './compB.ts'

// ....
const a = new CompA()
// ...
```
and my manifest.json
```json
{
  "include": [
    "$(MODDABLE)/examples/manifest_mod.json",
    "$(MODDABLE)/examples/manifest_typings.json"
  ],
  "config": {
    "version": 1.1
  },
  "modules": {
    "*": [
      "./src/*"
      , "./src/folderA/*"
    ]
  },
  "typescript": {
		"tsconfig": {
			"compilerOptions": {
				"types": [
					"$(MODDABLE)/xs/includes/xs",
					"$(MODDABLE)/typings/global"
				]
			}
		}
	}
}

```

And I found as long as the code is using CompA, the mod fails to load. If I move CompA.ts to the src/ folder and change the imports, things start working...

so... what's the deal?

**Comment by phoddie** [comment]
### Short answer
This "modules" section does what you want:

```json
"modules": {
"*": [
	"./src/*"
],
"./folderA/*": "./src/folderA/*"
```
### Examples

There are a couple of examples of this in the recent port I did of the node-chron package. Check out the manifests for [cron](https://github.com/phoddie/node-cron/blob/master/manifest_cron.json) and [luxon](https://github.com/phoddie/node-cron/blob/master/manifest_luxon.json). The read.me also talks about this topic specifically in the [manifest section](https://github.com/phoddie/node-cron/blob/master/README.md#manifest).

### Longer answer

In the modules section, the left side controls the module specifier that is used to access the module from JavaScript; the right side is the path the module (or modules) source code. When you write this...

    "*": [
      "./src/folderA/*"
    ],

... the "*" means to put the modules at the root (e.g. no sub-directories). If folderA contains "compA.ts" that will be available to import with the module specified "compA.ts".

Here you want to maintain the "folderA" part of the path but not "src". Instead of using "*" on the left, use the path you want in the module specifier:

	"./folderA/*": "./src/folderA/*"

Modules with source code in "./src/folderA" will be available from JavaScript module specifiers starting with "./folderA/".

> **Note**: Your example include the ".ts" extension when importing modules. That works. We generally omit that -- you can just import from `./folderA/compA`.
